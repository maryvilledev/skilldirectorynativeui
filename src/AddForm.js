import React, {Component} from "react"
import { ScrollView, Modal, Picker, Text, View, TextInput, Button, Switch, NativeMethodsMixin } from 'react-native'
import { scrollLayout, centerLayout, textStyles, horizontalLayout, formElement} from './Styles'
const Item = Picker.Item

const FORM_TYPE = {
  TEXT: 1,
  PICKER: 2,
  SWITCH: 3,
}

export {FORM_TYPE};

export default class AddForm extends Component {
  constructor(props) {
    super(props);

    this.state = {
      values: getDefaultValues(props.forms),
    }

    this.onSubmit = this.onSubmit.bind(this);
    this.onCancel = this.onCancel.bind(this);
    this.clearForms = this.clearForms.bind(this);
    this.onClose = props.onClose.bind(this);
  }

  componentWillReceiveProps(props) {
    this.state = {
      values: getDefaultValues(props.forms)
    }
  }

  update(index, value) {
    const values = this.state.values.slice()
    values[index] = value
    this.setState({
      values: values
    })
  }

  clearForms() {
    this.setState({
      values: getDefaultValues(this.props.forms),
    })
  }

  onSubmit() {
    this.props.onSubmit(this.state.values.slice());
    this.clearForms();
    this.onClose();
  }

  onCancel() {
    this.clearForms();
    this.onClose();
  }

  render() {
    let _scrollView: ScrollView;
    const onFocus = (y) => {_scrollView.scrollTo({y: y})};
    const onBlur = () => {_scrollView.scrollTo({y: 0})};
    const values = this.state.values.slice()
    const forms = this.props.forms.map((form, index) => <Form
      form={form}
      key={index}
      update={val => this.update(index, val)}
      value={values[index]}
      onFocus={onFocus}
      onBlur={onBlur}
    />)
    return (
      <View>
        <Modal
          animationType={"slide"}
          transparent={false}
          visible={this.props.visible}
          onRequestClose={this.onCancel}
        >
          <View style={{height: 20}} />
          {/* Strut to keep off top of screen */}
          <ScrollView
            ref={scrollView => {_scrollView = scrollView;}}
            style={scrollLayout}
          >
            {forms}
          </ScrollView>
          <View style={horizontalLayout}>
            <Button title="Submit" onPress={this.onSubmit} />
            <Button title="Cancel" onPress={this.onCancel} />
          </View>
        </Modal>
      </View>
    )
  }
}

const Form = (props) => {
  const form = props.form
  const update = props.update
  const value = props.value
  switch (form.type) {
    case FORM_TYPE.TEXT:
    const onFocus = props.onFocus;
    const onBlur = props.onBlur;
    let _y = 0;
      return (
      <View style={formElement} onLayout={ev => {_y = ev.nativeEvent.layout.y}}>
        <Text style={textStyles.large}>{`${form.label}: `}</Text>
        <TextInput
          style={{height: 40, borderColor: 'grey', borderWidth: 1}}
          onChangeText={update}
          value={value}
          multiline={form.multiline || false}
          onFocus={() => onFocus(_y)}
          onBlur={onBlur}
        />
      </View>
    );
    case FORM_TYPE.PICKER:
      return (
        <View style={formElement}>
          <Text style={textStyles.large}>{`${form.label}: `}</Text>
          <Picker
            onValueChange={update}
            selectedValue={value || form.items[0]}
          >
            {form.items.map((item, index) => <Item label={item} value={item} key={index}/>)}
          </Picker>
        </View>
      )
    case FORM_TYPE.SWITCH:
      return (
        <View style={formElement}>
          <Text style={textStyles.large}>{`${form.label}: `}</Text>
          <Switch
            onValueChange={update}
            value={value}
          />
        </View>
      )
  }
}

function getDefaultValues(forms) {
  let values = []
  for(form of forms){
    switch (form.type) {
      case FORM_TYPE.TEXT:
        values.push('')
        break;
      case FORM_TYPE.PICKER:
        values.push(form.items[0])
        break;
      case FORM_TYPE.SWITCH:
        values.push(false)
        break;
      default:
        values.push(null)
    }
  }
  return values
}
