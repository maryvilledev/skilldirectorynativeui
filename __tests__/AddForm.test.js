import 'react-native'
import React from 'react'
import AddForm, { FORM_TYPE } from './../src/AddForm.js'

import renderer from 'react-test-renderer'

const mockForms = [
  {type: FORM_TYPE.TEXT, label: 'Name', multiline: true},
  {type: FORM_TYPE.PICKER, label: 'Title', items: ['Destructor', 'Cheerleader', 'Rabbit']},
  {type: FORM_TYPE.SWITCH, label: 'PK Sensitive'}
]

describe('<AddForm />', () => {
  it('renders correctly', () => {
    const tree = renderer.create(
      <AddForm
        forms={mockForms}
        visible={true}
        onClose={jest.fn()}
        onSubmit={jest.fn()}
      />
    ).toJSON();
    expect(tree).toMatchSnapshot();
  })
})
