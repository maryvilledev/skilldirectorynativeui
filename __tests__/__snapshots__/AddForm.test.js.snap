exports[`<AddForm /> renders correctly 1`] = `
<View>
  <Modal
    animationType="slide"
    onRequestClose={[Function]}
    transparent={false}
    visible={true}>
    <View
      style={
        Object {
          "height": 20,
        }
      } />
    <ScrollView
      style={
        Object {
          "alignSelf": "stretch",
        }
      }>
      <View
        style={
          Object {
            "borderRadius": 3,
            "borderWidth": 0.5,
            "margin": 10,
            "marginVertical": 5,
            "overflow": "hidden",
          }
        }>
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
          style={
            Object {
              "fontSize": 30,
            }
          }>
          Name: 
        </Text>
        <TextInput
          multiline={true}
          onBlur={[Function]}
          onChangeText={[Function]}
          onFocus={[Function]}
          style={
            Object {
              "borderColor": "grey",
              "borderWidth": 1,
              "height": 40,
            }
          }
          value="" />
      </View>
      <View
        style={
          Object {
            "borderRadius": 3,
            "borderWidth": 0.5,
            "margin": 10,
            "marginVertical": 5,
            "overflow": "hidden",
          }
        }>
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
          style={
            Object {
              "fontSize": 30,
            }
          }>
          Title: 
        </Text>
        <View
          style={undefined}>
          <RCTPicker
            items={
              Array [
                Object {
                  "label": "Destructor",
                  "textColor": undefined,
                  "value": "Destructor",
                },
                Object {
                  "label": "Cheerleader",
                  "textColor": undefined,
                  "value": "Cheerleader",
                },
                Object {
                  "label": "Rabbit",
                  "textColor": undefined,
                  "value": "Rabbit",
                },
              ]
            }
            onChange={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            selectedIndex={0}
            style={
              Array [
                Object {
                  "height": 216,
                },
                undefined,
              ]
            } />
        </View>
      </View>
      <View
        style={
          Object {
            "borderRadius": 3,
            "borderWidth": 0.5,
            "margin": 10,
            "marginVertical": 5,
            "overflow": "hidden",
          }
        }>
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
          style={
            Object {
              "fontSize": 30,
            }
          }>
          PK Sensitive: 
        </Text>
        <RCTSwitch
          disabled={false}
          onChange={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          onValueChange={[Function]}
          style={
            Array [
              Object {
                "height": 31,
                "width": 51,
              },
              undefined,
            ]
          }
          value={false} />
      </View>
    </ScrollView>
    <View
      style={
        Object {
          "alignItems": "center",
          "alignSelf": "center",
          "flexDirection": "row",
          "justifyContent": "center",
        }
      }>
      <View
        accessibilityComponentType="button"
        accessibilityLabel={undefined}
        accessibilityTraits={
          Array [
            "button",
          ]
        }
        accessible={true}
        hitSlop={undefined}
        onLayout={undefined}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "opacity": 1,
          }
        }
        testID={undefined}>
        <View
          style={
            Array [
              Object {},
            ]
          }>
          <Text
            accessible={true}
            allowFontScaling={true}
            ellipsizeMode="tail"
            style={
              Array [
                Object {
                  "color": "#0C42FD",
                  "fontSize": 18,
                  "padding": 8,
                  "textAlign": "center",
                },
              ]
            }>
            Submit
          </Text>
        </View>
      </View>
      <View
        accessibilityComponentType="button"
        accessibilityLabel={undefined}
        accessibilityTraits={
          Array [
            "button",
          ]
        }
        accessible={true}
        hitSlop={undefined}
        onLayout={undefined}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "opacity": 1,
          }
        }
        testID={undefined}>
        <View
          style={
            Array [
              Object {},
            ]
          }>
          <Text
            accessible={true}
            allowFontScaling={true}
            ellipsizeMode="tail"
            style={
              Array [
                Object {
                  "color": "#0C42FD",
                  "fontSize": 18,
                  "padding": 8,
                  "textAlign": "center",
                },
              ]
            }>
            Cancel
          </Text>
        </View>
      </View>
    </View>
  </Modal>
</View>
`;
